ofrak_core/ofrak/gui/script_builder.py

```diff
@@ -0,0 +1,280 @@
+        )
+        await self._add_variable_to_session(resource, name)
+        return name
+
+    async def add_action(
+        self,
+        resource: Resource,
+        action: str,
+        action_type: ActionType,
+    ) -> None:
+        """
+        :param action:
+        :param action_type:
+        """
+        var_name = await self.add_variable(resource)
+        qualified_action = action.replace("$resource", var_name)
+        await self._add_action_to_session(resource, qualified_action, action_type)
+
+    async def _add_action_to_session(self, resource, action, action_type):
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        # TODO: actions are duplicated if page is refreshed, is this reasonable?
+        session.hashed_actions[session.actions_counter] = ScriptAction(action_type, action)
+        session.actions_counter += 1
+
+    async def _add_variable_to_session(self, resource: Resource, var_name: str):
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource.get_id())
+        session.variable_mapping[resource.get_id()] = var_name
+
+    async def _get_variable_from_session(self, resource: Resource):
+        root_resource = await self._get_root_resource(resource)
+        return self.script_sessions[root_resource.get_id()].variable_mapping[resource.get_id()]
+
+    async def _var_exists(self, resource: Resource):
+        root_resource = await self._get_root_resource(resource)
+        session = self._get_session(root_resource)
+        return resource.get_id() in session.variable_mapping
+
+    def delete_action(self, resource_id: bytes, action: str) -> None:
+        """
+        :param action:
+        """
+        # TODO: do we really need to delete an action from the script?
+        for key, script_action in self.script_sessions[resource_id].hashed_actions.items():
+            if script_action.action == action:
+                del self.script_sessions[resource_id].hashed_actions[key]
+
+    def _get_session(self, resource_id: bytes) -> ScriptSession:
+        session = self.script_sessions.get(resource_id, None)
+        if session is None:
+            session = ScriptSession()
+            self.script_sessions[resource_id] = session
+
+        return session
+
+    async def get_script(self, resource: Resource) -> str:
+        """
+        :return script:
+        """
+        root_resource = await self._get_root_resource(resource)
+        return self._get_script(root_resource.get_id())
+
+    def _get_script(self, resource_id: bytes, target_type: Optional[ActionType] = None) -> str:
+        script = []
+        script.append(self.script_sessions[resource_id].boilerplate_header)
+        for script_action in self.script_sessions[resource_id].hashed_actions.values():
+            if target_type is None or target_type == script_action.action_type:
```

Comment:
Seems like `UNDEF` actions should be allowed here too, since that includes things like assigning values to variables, variables which might be used in later e.g. `UNPACK` actions (or whichever match the `target_type`). `target_type` seems more like it should be a set of allowed types, and then `get_all_of_type` determines what that set is.

Though tbh `get_all_of_type` is unused and I'm not sure it belongs in MVP anyway, so maybe `get_all_of_type` should just raise `NotImplementedError` and we don't try to filter by action type in this function.